filters/auth/jwt_validation.go

```diff
@@ -0,0 +1,301 @@
+	}
+
+	// inject additional headers from the access token for upstream applications
+	if len(sargs) > 2 && sargs[2] != "" {
+		f.upstreamHeaders = make(map[string]string)
+
+		for _, header := range strings.Split(sargs[2], " ") {
+			sl := strings.SplitN(header, ":", 2)
+			if len(sl) != 2 || sl[0] == "" || sl[1] == "" {
+				return nil, fmt.Errorf("%w: malformatted filter for upstream headers %s", filters.ErrInvalidFilterParameters, sl)
+			}
+			f.upstreamHeaders[sl[0]] = sl[1]
+		}
+		log.Debugf("Upstream Headers: %v", f.upstreamHeaders)
+	}
+
+	return f, nil
+}
+
+// String prints nicely the jwtValidationFilter configuration based on the
+// configuration and check used.
+func (f *jwtValidationFilter) String() string {
+	return fmt.Sprintf("%s(%s)", JwtValidationName, strings.Join(f.claims, ","))
+}
+
+func (f *jwtValidationFilter) validateAnyClaims(token jwt.Token) bool {
+	for _, wantedClaim := range f.claims {
+		if claims, ok := token.Claims.(jwt.MapClaims); ok {
+			if _, ok2 := claims[wantedClaim]; ok2 {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+func (f *jwtValidationFilter) Request(ctx filters.FilterContext) {
+	r := ctx.Request()
+
+	var info jwt.Token
+	infoTemp, ok := ctx.StateBag()[oidcClaimsCacheKey]
+	if !ok {
+		token, ok := getToken(r)
+		if !ok || token == "" {
+			unauthorized(ctx, "", missingToken, "", "")
+			return
+		}
+
+		parsedToken, err := jwt.Parse(token, func(token *jwt.Token) (interface{}, error) {
+			kid := token.Header["kid"].(string)
+
+			m.Lock()
+			defer m.Unlock()
+
+			if rsakeys[kid] == nil {
+				return nil, fmt.Errorf("not able to find public key for kid %s", kid)
+			}
+
+			return rsakeys[kid], nil
+		})
+		if err != nil {
+			log.Errorf("Error while parsing jwt token : %v.", err)
+			unauthorized(ctx, "", "error parsing jwt token", "", "")
+			return
+		} else if !parsedToken.Valid {
+			log.Errorf("Invalid token")
+			unauthorized(ctx, "", "Invalid token", "", "")
+			return
+		} else if parsedToken.Header["alg"] == nil {
+			log.Errorf("alg must be defined")
+			unauthorized(ctx, "", "alg must be defined", "", "")
+			return
+		}
+
+		info = *parsedToken
+		/*err = json.Unmarshal([]byte(parsedToken.Raw), &info)
+		if err != nil {
+			log.Errorf("Error while pasing jwt token : %v.", err)
+			unauthorized(ctx, "", "error parsing jwt token", f.authClient.url.Hostname(), "")
+			return*/
+		//}
+	} else {
+		info = infoTemp.(jwt.Token)
+	}
+
+	/*sub, err := info.Sub()
+	if err != nil {
+		if err != errInvalidTokenintrospectionData {
+			log.Errorf("Error while reading token: %v.", err)
+		}
+
+		unauthorized(ctx, sub, invalidSub, f.authClient.url.Hostname(), "")
+		return
+	}
+
+	if !info.Active() {
+		unauthorized(ctx, sub, inactiveToken, f.authClient.url.Hostname(), "")
+		return
+	}*/
+
+	sub := info.Claims.(jwt.MapClaims)["sub"].(string)
+
+	var allowed = f.validateAnyClaims(info)
```

Comment:
I see there is a family of `oauthTokenintrospection*` filters that validate any, all claims or key-values and I am wondering would we need to support the similar set of validations? If so I would prefer that we reuse validation code instead of copy-pasting.

Also would it be possible to completely remove claims validation here and rely on the `oidcClaimsQuery()` instead reducing this filter scope only to the JWT signature validation?

Reply:
This make sense from one side. From another side I was trying to keep the similarity to oidc flow where filter does check the claims